File: app/src/main/java/com/braingroom/tutor/viewmodel/activity/MyClassesViewModel.java
package com.braingroom.tutor.viewmodel.activity;

import android.util.Log;

import com.braingroom.tutor.R;
import com.braingroom.tutor.model.data.ListDialogData;
import com.braingroom.tutor.model.req.ClassListReq.Snippet;
import com.braingroom.tutor.model.resp.ClassListResp;
import com.braingroom.tutor.view.adapters.ViewProvider;
import com.braingroom.tutor.viewmodel.ViewModel;
import com.braingroom.tutor.viewmodel.item.ClassListItemViewModel;
import com.braingroom.tutor.viewmodel.item.ListDialogViewModel;
import com.braingroom.tutor.viewmodel.item.LoadingViewModel;
import com.braingroom.tutor.viewmodel.item.NotifyDataSetChanged;
import com.braingroom.tutor.viewmodel.item.RemoveLoadingViewModel;

import java.util.HashMap;
import java.util.LinkedHashMap;

import io.reactivex.Observable;

import static com.braingroom.tutor.utils.CommonUtilsKt.toObservable;


public class MyClassesViewModel extends ViewModel {
    public final ListDialogViewModel classType;
    public final ListDialogViewModel classStatus;
    public final Snippet snippet = new Snippet(true, true, getUserId());
    private int currentPageNumber = 0;
    public ViewProvider view = vm -> {
        if (vm != null) {
            if (vm instanceof ClassListItemViewModel)
                return R.layout.item_class_list;
            if (vm instanceof LoadingViewModel)
                return R.layout.item_loading_class_list;
            else
                try {
                    throw new NoSuchFieldException();
                } catch (Exception e) {
                    throw new Error();
                }
        }
        throw new NullPointerException();
    };

    public MyClassesViewModel() {


        LinkedHashMap<String, Integer> classTypeData = new LinkedHashMap<>();
        classTypeData.put("Batch", 1);
        classTypeData.put("Flexible", 0);

        LinkedHashMap<String, Integer> classStatusData = new LinkedHashMap<>();
        classStatusData.put("Expired", 1);
        classStatusData.put("Ongoing", 0);

        classType = new ListDialogViewModel("Class Type", Observable.just(new ListDialogData(classTypeData)), new HashMap<>(), false, selectedItems -> {
            if (selectedItems.values().iterator().hasNext()) {
                snippet.setBatch(selectedItems.values().iterator().next());
                reset();
                Log.d(getTAG(), "classType selectedItems items : " + selectedItems.values());
            }
        }, "", null);

        classStatus = new ListDialogViewModel("Class Status", Observable.just(new ListDialogData(classStatusData)), new HashMap<>(), false, selectedItems -> {
            if (selectedItems.values().iterator().hasNext()) {
                snippet.setExpired(selectedItems.values().iterator().next());
                reset();
                Log.d(getTAG(), "classStatus selectedItems items : " + selectedItems.values());
            }
        }, "", null);
        toObservable(getCallAgain()).subscribe(integer -> getApiService().getAllClasses(snippet, currentPageNumber).doOnSubscribe(disposable -> {
            for (int i = 0; i < 5; i++)
                getItem().onNext(new LoadingViewModel());
            getItem().onNext(new NotifyDataSetChanged());
            getCompositeDisposable().add(disposable);
            Log.d(getTAG(), "Added  Loading Items");
        }).
                doOnComplete(() -> {
                    Log.d(getTAG(), "Updating UI");
                    getItem().onNext(new NotifyDataSetChanged());
                }).
                subscribe(resp ->
                {

                    getItem().onNext(new RemoveLoadingViewModel());
                    Log.d(getTAG(), "Removed Loading Items");
                    if (resp.getResCode()) {
                        for (ClassListResp.Snippet snippet : resp.getData())
                            getItem().onNext(new ClassListItemViewModel(snippet));
                        Log.d(getTAG(), "Added Actual items");
                        currentPageNumber++;
                    }

                }, throwable -> {
                    Log.d(getTAG(), throwable.getMessage());
                    throwable.printStackTrace();
                }));
    }

    private void reset() {
        currentPageNumber = 0;
        getCallAgain().set(getCallAgain().get() + 1);
    }
}

File: app/src/main/java/com/braingroom/tutor/viewmodel/activity/MyClassesViewModel.kt
package com.braingroom.tutor.viewmodel.activity

import android.os.Bundle
import android.util.Log

import com.braingroom.tutor.R
import com.braingroom.tutor.common.modules.HelperFactory
import com.braingroom.tutor.model.data.ListDialogData
import com.braingroom.tutor.model.req.ClassListReq.Snippet
import com.braingroom.tutor.model.resp.BaseResp
import com.braingroom.tutor.model.resp.ClassListResp
import com.braingroom.tutor.view.adapters.ViewProvider
import com.braingroom.tutor.viewmodel.ViewModel

import io.reactivex.Observable

import com.braingroom.tutor.utils.toObservable
import com.braingroom.tutor.view.activity.ClassDetailActivity
import com.braingroom.tutor.viewmodel.item.*
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Action
import io.reactivex.functions.Consumer
import java.util.*


class MyClassesViewModel(helperFactory: HelperFactory) : ViewModel(helperFactory) {
    val classType: ListDialogViewModel
    val classStatus: ListDialogViewModel
    val snippet = Snippet(userId)
    val viewProvider: ViewProvider by lazy {
        object : ViewProvider {
            override fun getView(vm: RecyclerViewItem?): Int {
                return when (vm) {
                    is ClassListItemViewModel -> R.layout.item_class_list
                    is LoadingViewModel -> R.layout.item_loading_class_list
                    null -> throw NullPointerException()
                    else -> throw NoSuchFieldError()
                }
            }
        }
    }

    init {
        object : Consumer<MyProfileItem> {
            override fun accept(t: MyProfileItem?) {
                TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
            }
        }
        val classTypeData = LinkedHashMap<Int, String>()
        classTypeData.put(1, "Batch")
        classTypeData.put(2, "Flexible")


        val classStatusData = LinkedHashMap<Int, String>()
        classStatusData.put(1, "Expired")
        classStatusData.put(2, "Ongoing")

        classType = ListDialogViewModel(helperFactory, "Class Type", Observable.just(ListDialogData(classTypeData)), HashMap(), false, Consumer {
            it.keys.forEach { snippet.setBatch(it) }
            reset()
        }, "", null)

        classStatus = ListDialogViewModel(helperFactory, "Class Status", Observable.just(ListDialogData(classStatusData)), HashMap(), false, Consumer {
            it.keys.forEach { snippet.setExpired(it) }
            reset()
        }, "", null)
        toObservable(callAgain).filter { pageNumber > -1 && !paginationInProgress }.doOnSubscribe { compositeDisposable.add(it) }.
                subscribe {
                    apiService.getAllClasses(snippet, pageNumber).
                            map(this::respToViewModelList).
                            doOnSubscribe(this::addLoadingItems).
                            subscribe(this::addActualItems, this::handleError)
                }

    }

    private fun reset() {
        pageNumber = 1
        paginationInProgress = false
        item.onNext(RefreshViewModel())
        callAgain.set(callAgain.get() + 1)
    }

    private fun respToViewModelList(resp: ClassListResp): List<RecyclerViewItem> = when {
        resp.resCode -> {
            pageNumber++
            resp.data.map {
                ClassListItemViewModel(it, Action {
                    apiService.getClassDetail(it.classId.toString()).doOnSubscribe {
                        compositeDisposable.add(it)
                        messageHelper.showProgressDialog("Wait", "Loading")
                    }.doOnError(this::handleError).doOnError { messageHelper.dismissActiveProgress() }.observeOn(AndroidSchedulers.mainThread()).
                            subscribe {
                                if (it.resCode) {
                                    messageHelper.dismissActiveProgress()
                                    val data = Bundle()
                                    data.putSerializable("classData", it.data)
                                    navigator.navigateActivity(ClassDetailActivity::class.java, data)
                                } else messageHelper.showMessage(it.resMsg)
                            }

                })
            }
        }
        pageNumber == 1 -> {
            pageNumber = -1
            Collections.singletonList(EmptyItemViewModel("", R.drawable.ic_no_post_64dp, "No Classes"))
        }
        else -> {
            pageNumber = -1
            ArrayList()
        }
    }


    private fun addActualItems(viewModelList: List<RecyclerViewItem>) {
        item.onNext(RemoveLoadingViewModel())
        item.onNext(NotifyDataSetChanged())
        viewModelList.forEach { item.onNext(it) }
        item.onNext(NotifyDataSetChanged())
        paginationInProgress = false
    }

    private fun navigateToClassDetail() {
        apiService.getClassDetail("743").doOnSubscribe {
            compositeDisposable.add(it)
            messageHelper.showProgressDialog("Wait", "Loading")
        }.doOnError(this::handleError).doOnError { messageHelper.dismissActiveProgress() }.observeOn(AndroidSchedulers.mainThread()).
                subscribe {
                    if (it.resCode) {
                        messageHelper.dismissActiveProgress()
                        val data = Bundle()
                        data.putSerializable("classData", it.data)
                        navigator.navigateActivity(ClassDetailActivity::class.java, data)
                    } else messageHelper.showMessage(it.resMsg)
                }
    }

}

